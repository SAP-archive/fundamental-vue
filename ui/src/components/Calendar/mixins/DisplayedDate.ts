import { monthFromDate, Month } from "./../util";
import Vue from "vue";
import { PropValidator } from "vue/types/options";

const defaultDayNames = ["Su", "M", "Tu", "W", "Th", "F", "S"];
const defaultMonthNames = [
  "Jan.",
  "Feb.",
  "Mar.",
  "Apr.",
  "May",
  "Jun.",
  "Jul",
  "Aug.",
  "Sep.",
  "Oct.",
  "Nov.",
  "Dec."
];

export default Vue.extend({
  props: {
    defaultValue: {
      type: [Date, String, Number] as any,
      default: Date.now()
    } as PropValidator<Date | string | number>,
    firstDayOfWeek: { type: Number, default: 0 } as PropValidator<number>,
    dayNames: { type: Array, default: () => defaultDayNames } as PropValidator<
      string[]
    >,
    monthNames: {
      type: Array,
      default: () => defaultMonthNames
    } as PropValidator<string[]>
  },
  data() {
    return {
      displayedDate:
        typeof this.defaultValue === "string" ||
        typeof this.defaultValue === "number"
          ? new Date(this.defaultValue)
          : this.defaultValue
    };
  },
  computed: {
    month(): Month {
      const date = this.displayedDate;
      return monthFromDate(date, { firstDayOfWeek: this.firstDayOfWeek });
    },
    adjustedDayNames(): string[] {
      return this.dayNames
        .concat(this.dayNames)
        .slice(this.firstDayOfWeek, this.firstDayOfWeek + 7);
    },
    displayedYear: {
      get(): number {
        return this.displayedDate.getFullYear();
      },
      set(year: number) {
        const copy = new Date(this.displayedDate);
        copy.setFullYear(year);
        this.displayedDate = copy;
      }
    },
    displayedMonth: {
      set(month: number) {
        const copy = new Date(this.displayedDate);
        copy.setMonth(month);
        this.displayedDate = copy;
      },
      get(): number {
        const date = this.displayedDate;
        return date.getMonth();
      }
    }
  }
});
