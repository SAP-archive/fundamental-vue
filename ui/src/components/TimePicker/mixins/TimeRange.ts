import { TimeType } from "./../Time/TimeType";

import Vue from "vue";

const makeDefaultRange = () => ({
  hour24: {
    min: "00",
    max: "23"
  },
  hour12: {
    min: "01",
    max: "12"
  },
  minute: {
    min: "00",
    max: "59"
  },
  second: {
    min: "00",
    max: "59"
  },
  meridian: {
    min: "am",
    max: "pm"
  }
});

type RangeType = ReturnType<typeof makeDefaultRange>;

export default Vue.extend({
  methods: {
    checkValueRange(timeValue: string | number, timeType: TimeType): boolean {
      return (
        !isNaN(Number(timeValue)) &&
        (Number(timeValue) >= Number(this.range[timeType].min) &&
          Number(timeValue) <= Number(this.range[timeType].max))
      );
    }
  },
  data() {
    return {
      range: makeDefaultRange() as RangeType
    };
  }
});
