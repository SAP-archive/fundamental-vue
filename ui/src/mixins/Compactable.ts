export default {
  inject: {
    compactableContainer: { default: { compact: false } }
  },
  computed: {
    enclosedByCompactContainer(): boolean {
      // @ts-ignore
      return this.compactableContainer.compact;
    },
    compact() {
      return this.enclosedByCompactContainer;
    }
  }
};
