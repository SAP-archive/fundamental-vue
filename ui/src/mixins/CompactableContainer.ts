import Vue from "vue";

export default {
  beforeCreate() {
    // @ts-ignore
    this.compactableContainer = Vue.observable({ compact: false });
  },
  provide(): object {
    // @ts-ignore
    return { compactableContainer: this.compactableContainer };
  },
  watch: {
    compact: {
      immediate: true,
      handler(newCompact: boolean) {
        // @ts-ignore
        this.compactableContainer.compact = newCompact;
      }
    }
  },
  props: {
    compact: {
      type: Boolean,
      default: false
    }
  }
};
