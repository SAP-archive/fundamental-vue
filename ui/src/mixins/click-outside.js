import { shortUuid } from "./../lib";

export default {
  data() {
    return {
      clickOutsideContext: shortUuid()
    };
  }
};
