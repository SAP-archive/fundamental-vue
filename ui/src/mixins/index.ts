export { default as Uid } from "./Uid";
export { default as Icon } from "./Icon";
export { default as mixins } from "./typesafeMixins";
export { default as withTargetLocation } from "./withTargetLocation";
// @ts-ignore
export { default as Compactable } from "./Compactable";
// @ts-ignore
export { default as CompactableContainer } from "./CompactableContainer";
export { default as FocusTrap } from "./FocusTrap";
