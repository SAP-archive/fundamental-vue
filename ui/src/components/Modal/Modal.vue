<template>
  <transition name="fade">
    <div
      v-show="isActive"
      class="fd-ui__overlay fd-overlay fd-overlay--modal"
      :aria-hidden="!isActive"
    >
      <ClickAwayContainer
        @clickOutside="clickOutside"
        class="fd-modal"
        :tabindex="isActive ? -1 : 0"
        :active="isActive"
      >
        <div class="fd-modal__content" role="document">
          <!-- HEADER -->
          <div class="fd-modal__header">
            <h1 class="fd-modal__title">{{ title }}</h1>
            <Button
              class="fd-modal__close"
              styling="light"
              @click="close"
              aria-label="close"
            />
          </div>

          <!-- BODY -->
          <div class="fd-modal__body">
            <slot />
          </div>

          <!-- FOOTER -->
          <footer
            v-if="$slots.footer != null || $slots.actions != null"
            class="fd-modal__footer"
          >
            <slot name="footer" />
            <div v-if="$slots.actions" class="fd-modal__actions">
              <slot name="actions" />
            </div>
          </footer>
        </div>
      </ClickAwayContainer>
    </div>
  </transition>
</template>

<script lang="ts">
// Use these types in order to cast your props. Delete if not needed.
// import { PropValidator } from "vue/types/options";
// import { Prop } from "vue/types/options";
import ClickAwayContainer from "@/components/ClickAwayContainer";
import { Button } from "@/components/Button";
import { FocusTrap, mixins } from "@/mixins";

export default mixins(FocusTrap).extend({
  name: "FdModal",
  mounted() {
    document.body.appendChild(this.$el);
    this.initializeFocusTrap(this.$el, {
      onDeactivate: this.close,
      initialFocus: ".fd-modal"
    });
  },
  destroyed() {
    const el = this.$el;
    if (el != null) {
      const parent = el.parentNode;
      if (parent != null) {
        parent.removeChild(el);
      }
    }
  },
  updated() {
    if (this.isActive) {
      this.activateFocusTrap();
    } else {
      this.deactivateFocusTrap();
    }
  },
  props: {
    active: { type: Boolean, default: false },
    title: { type: String, default: null }
  },
  components: { Button, ClickAwayContainer },
  watch: {
    active: {
      immediate: true,
      handler(newIsActive: boolean) {
        this.isActive = newIsActive;
      }
    }
  },
  methods: {
    clickOutside() {
      if (this.isActive) {
        this.close();
      }
    },
    close() {
      this.$emit("close");
      this.$emit("update:active", false);
    }
  },
  data() {
    return {
      isActive: this.active
    };
  }
});
</script>
