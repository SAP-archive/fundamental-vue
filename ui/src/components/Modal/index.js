import Modal from "./Modal.vue";
import { pluginify } from "@/util";
export default pluginify(Modal);
export { Modal };
