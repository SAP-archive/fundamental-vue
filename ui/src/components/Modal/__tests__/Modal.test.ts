import Vue from "vue";
import { mount, Wrapper, shallowMount } from "@vue/test-utils";
import Modal from "../Modal.vue";
import { FocusTrap } from "@/mixins";
import createFocusTrap, { Options } from "focus-trap";

jest.mock("focus-trap");
interface ModalWrapper extends Vue {
  clickOutside(): void;
  close(): void;
  activateFocusTrap(): void;
  deactivateFocusTrap(): void;
  initializeFocusTrap(element: Element, options?: Options): void;
}

describe("Modal", () => {
  it("renders correctly", () => {
    const wrapper = mount(Modal, { mixins: [FocusTrap] });
    expect(wrapper.element).toMatchSnapshot();
  });

  it("should update tabindex when opened", () => {
    const wrapper = shallowMount(Modal, {
      mixins: [FocusTrap],
      propsData: { active: true }
    });

    expect(wrapper.find(".fd-modal").attributes("tabindex")).toBe("-1");
  });

  it("should update tabindex when closed", () => {
    const wrapper = shallowMount(Modal, {
      mixins: [FocusTrap],
      propsData: { active: false }
    });

    expect(wrapper.find(".fd-modal").attributes("tabindex")).toBe("0");
  });

  it("should initialize focus trap while mounting", () => {
    const initializeFocusTrapMock = jest.fn();

    const wrapper: Wrapper<ModalWrapper> = shallowMount(Modal, {
      mixins: [FocusTrap],
      methods: {
        initializeFocusTrap: initializeFocusTrapMock
      }
    }) as Wrapper<ModalWrapper>;

    expect(initializeFocusTrapMock).toHaveBeenCalledWith(wrapper.element, {
      initialFocus: ".fd-modal",
      onDeactivate: wrapper.vm.close
    });
  });

  describe("methods", () => {
    describe("initializeFocusTrap", () => {
      it("should create trap on given HTML element", () => {
        const mockElement = document.createElement("div");
        const wrapper = shallowMount(Modal, {
          mixins: [FocusTrap]
        }) as Wrapper<ModalWrapper>;

        wrapper.vm.initializeFocusTrap(mockElement, {});

        expect(createFocusTrap).toHaveBeenCalledWith(mockElement, {});
      });
    });

    describe("activateFocusTrap", () => {
      it("should activate focus trap", () => {
        const focusTrapMock = {
          activate: jest.fn(),
          deactivate: jest.fn()
        };
        (<jest.Mock>createFocusTrap).mockReturnValue(focusTrapMock);
        const wrapper = shallowMount(Modal, {
          mixins: [FocusTrap]
        }) as Wrapper<ModalWrapper>;

        wrapper.vm.activateFocusTrap();

        expect(focusTrapMock.activate).toHaveBeenCalled();
      });
    });

    describe("deactivateFocusTrap", () => {
      it("should deactivate focus trap", () => {
        const focusTrapMock = {
          activate: jest.fn(),
          deactivate: jest.fn()
        };
        (<jest.Mock>createFocusTrap).mockReturnValue(focusTrapMock);
        const wrapper = shallowMount(Modal, {
          mixins: [FocusTrap]
        }) as Wrapper<ModalWrapper>;

        wrapper.vm.activateFocusTrap();

        expect(focusTrapMock.activate).toHaveBeenCalled();
      });
    });
  });
});
