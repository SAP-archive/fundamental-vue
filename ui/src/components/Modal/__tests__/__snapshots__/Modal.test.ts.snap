// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal renders correctly 1`] = `
<div
  aria-hidden="true"
  class="fd-ui__overlay fd-overlay fd-overlay--modal"
  name="fade"
  style="display: none;"
>
  <div
    class="fd-modal"
    tabindex="0"
  >
    <div
      class="fd-modal__content"
      role="document"
    >
      <div
        class="fd-modal__header"
      >
        <h1
          class="fd-modal__title"
        >
          
        </h1>
         
        <button
          aria-label="close"
          class="fd-modal__close fd-button--light"
        />
      </div>
       
      <div
        class="fd-modal__body"
      />
       
      <!---->
    </div>
  </div>
</div>
`;
